// angular-paginator/esm5/app/angular-paginator/pipes/angular-paginator.pipe.js
import * as tslib_1 from "tslib";
import { Pipe } from '@angular/core';
import { AngularPaginatorService } from '../services/angular-paginator.service';
/**
 * Pagination Pipe which paginates the array for the provided arguments such as `itemsPerPage` and `currentPage`
 */
var AngularPaginatorPipe = /** @class */ (function () {
    /**
     *
     * @param angularPaginatorService service for angular pagiantor
     */
    function AngularPaginatorPipe(angularPaginatorService) {
        this.angularPaginatorService = angularPaginatorService;
    }
    /**
     * this pipes paginates the array for the provided arguments such as `itemsPerPage` and `currentPage`
     *
     * @param array input array for which the manipulation happens
     * @param args input arguments for the paginator pipe
     */
    AngularPaginatorPipe.prototype.transform = function (array, args) {
        var instance = this.createInstance(array, args);
        // create pagination information
        this.angularPaginatorService.registerInstance(instance);
        // set the slicing range
        var start = (instance.currentPage - 1) * instance.itemsPerPage;
        var end = instance.currentPage * instance.itemsPerPage;
        return array.slice(start, end);
    };
    /**
     * create an instance for each pipe
     *
     * @param array input array for which the manipulation happens
     * @param args input arguments for the paginator pipe
     */
    AngularPaginatorPipe.prototype.createInstance = function (array, args) {
        return {
            id: args.id ? args.id : this.angularPaginatorService.id,
            currentPage: args.currentPage ? args.currentPage : 1,
            itemsPerPage: args.itemsPerPage ? args.itemsPerPage : 10,
            totalItems: array.length
        };
    };
    AngularPaginatorPipe = tslib_1.__decorate([
        Pipe({
            name: 'angularPaginator'
        }),
        tslib_1.__metadata("design:paramtypes", [AngularPaginatorService])
    ], AngularPaginatorPipe);
    return AngularPaginatorPipe;
}());
export { AngularPaginatorPipe };
